### PR TITLE
Add recorder for OperationModel

### PR DESCRIPTION
**Pull request by daniel-thom**

### CODE REVIEW

File: src/initial_conditions/update_initial_conditions.jl [hunk 1]
```diff
@@ -1,4 +1,4 @@
-function update_initial_conditions(
+function update_initial_conditions!(
```
Comment (daniel-thom) [inline src/initial_conditions/update_initial_conditions.jl:1]: Not sure if no `!` was intentional, but it felt wrong to me. Tell me if I should revert these changes.

File: src/initial_conditions/update_initial_conditions.jl [hunk 2]
```diff
@@ -130,9 +130,16 @@ function update_initial_conditions!(
...
     interval = get_interval(model.internal.store_parameters)
     ini_conditions_vector = get_initial_condition(container, key)
-    update_initial_conditions(ini_conditions_vector, model.store, interval)
+    timestamp = get_initial_time(model) + get_execution_count(model) * get_resolution(model)
+    update_initial_conditions!(ini_conditions_vector, model.store, interval)
+    for (i, initial_condition) in enumerate(ini_conditions_vector)
+        IS.@record :simulation InitialConditionUpdateEvent(
+            timestamp,
+            initial_condition,
+            1,  # TODO DT
```
Comment (daniel-thom) [inline src/initial_conditions/update_initial_conditions.jl:142]: We probably need this because the event is shared with simulations.

File: src/utils/recorder_events.jl [hunk 3]
```diff
@@ -59,24 +59,62 @@ end
...
+    common::IS.RecorderEventCommon
+    category::String
+    simulation_time::Dates.DateTime
+    parameter_type::String
+    device_name::String
+    previous_value::Float64
+    val::Float64
+    problem_number::Int
+    source::Int
+end
```
Comment (daniel-thom) [inline src/utils/recorder_events.jl:95]: This is not correct and TBD. Let me know what fields should be recorded for Parameter Updates. The implementation is as of now unclear.

File: src/operation/decision_model.jl [hunk 4]
```diff
@@ -480,7 +488,7 @@ function _write_model_dual_results!(
             key,
             timestamp,
             constraint,
-            [encode_key(key)],  # TODO DT: this doesn't seem right
+            [encode_key(key)],  # TODO DT: is this what the columns should be?
```
Comment (daniel-thom) [inline src/operation/decision_model.jl:491]: This code is only called in the Simulation execution path. The original code used the name (Symbol) of the dual. Is this going to be correct?

File: src/utils/recorder_events.jl [hunk 5]
```diff
@@ -48,35 +48,59 @@ end
...
         problem_number,
     )
 end
 
+struct ParameterUpdateEvent <: IS.AbstractRecorderEvent
+    common::IS.RecorderEventCommon
+    execution_timestamp::Dates.DateTime
+    parameter_type::String
+    device_type::String
+    name::String
```
Comment (daniel-thom) [inline src/utils/recorder_events.jl:83]: This field name may need to change. It is only valid if all ParameterAttributes subtypes will define `name`. We could call it something generic like `meta` or `tag`.

File: src/operation/decision_model.jl (hunk 4 above)
Comment (jd-lara) [reply src/operation/decision_model.jl:491]: It shouldn't be needed anymore since the duals are not related to the constraints that have their own column names. I think we need to test this in an emulation test that asks for duals.

File: src/utils/recorder_events.jl (hunk 5 above)
Comment (jd-lara) [reply src/utils/recorder_events.jl]: `timeseries_name`?

**Review by jd-lara — APPROVED** [review]

File: src/operation/decision_model.jl (hunk 4 above)
Comment (daniel-thom) [reply src/operation/decision_model.jl:491]: If you're saying that that might expose a problem in the EmulationModel/InMemoryModelStore code, I agree.

File: src/utils/recorder_events.jl (hunk 5 above)
Comment (daniel-thom) [reply src/utils/recorder_events.jl]: Note that I changed the field from `name` to `tag`. The base recorder event already uses `name` for the event type.

Using `time_series_name` implies that the event is only for TimeSeriesParameters, but that isn't right case, right? Won't we soon add events for other parameter types? I would be OK changing this event to be TimeSeriesParameterUpdateEvent. A consequence of doing that is that we would not be able to print all parameter update events in the same table.
